pkg/p2p/kadcast/utils.go

```diff
@@ -247,7 +248,42 @@ func generateRandomDelegates(beta uint8, in []encoding.PeerInfo, out *[]encoding
 	in[ind] = in[len(in)-1]
 	in = in[:len(in)-1]
 
-	return generateRandomDelegates(beta, in, out)
+	return getRandDelegates(beta, in, out)
+}
+
+// getRandDelegatesByShuffle selects n random and distinct items from `in` and
+// return them into `out` slice (no duplicates).
+// For the performance benefit, this is based on pseudo-randomizes.
+func getRandDelegatesByShuffle(beta uint8, in []encoding.PeerInfo) []encoding.PeerInfo {
+	mathrand.Seed(time.Now().UnixNano())
+	mathrand.Shuffle(len(in), func(i, j int) { in[i], in[j] = in[j], in[i] })
+
+	out := make([]encoding.PeerInfo, 0, beta)
+
+	for i := 0; i < int(beta); i++ {
+		if i >= len(in) {
+			break
+		}
+
+		out = append(out, in[i])
+	}
+
+	return out
+}
+
+// getNoRandDelegates selects n in a row items from `in` without any rand function.
+// Used for benchmarking.
+func getNoRandDelegates(beta uint8, in []encoding.PeerInfo) []encoding.PeerInfo {
```

Comment:
Probably amend that comment, since you also use it in the actual code - I was under the impression this was for testing only.

Reply:
This should be temporary. Before opening the PR for review, I'll probably remove it.